signer/src/schnorr.rs

```diff
@@ -83,23 +84,41 @@ impl<H: 'static + Hashable> Signer<H> for Schnorr<H> {
 
         rv.y.into_repr().is_even() && rv.x == sig.rx
     }
+
+    fn reset(&mut self) {
+        self.hasher.reset();
+    }
 }
 
 pub(crate) fn create_legacy<H: 'static + Hashable>(domain_param: H::D) -> impl Signer<H> {
-    Schnorr::<H> {
-        hasher: Box::new(mina_hasher::create_legacy::<Message<H>>(domain_param)),
+    Schnorr::new(
+        mina_hasher::create_legacy::<Message<H>>(domain_param),
         domain_param,
-    }
+    )
 }
 
 pub(crate) fn create_kimchi<H: 'static + Hashable>(domain_param: H::D) -> impl Signer<H> {
-    Schnorr::<H> {
-        hasher: Box::new(mina_hasher::create_kimchi::<Message<H>>(domain_param)),
+    Schnorr::new(
+        mina_hasher::create_kimchi::<Message<H>>(domain_param),
         domain_param,
-    }
+    )
 }
 
-impl<H: 'static + Hashable> Schnorr<H> {
+impl<T: Hasher<Message<H>>, H: 'static + Hashable> Schnorr<T, H> {
+    /// Creates a schnorr instance
+    pub fn new(hasher: T, domain_param: H::D) -> Self {
+        Self {
+            hasher,
+            domain_param,
+        }
+    }
+
+    /// Initiates inner hasher with the given domain param
+    pub fn init_domain_param(&mut self, domain_param: H::D) {
```

Comment:
This API is for supporting pooling, I believe it will be no longer relevant if #675  and #677  are accepted